hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/command/payload/ExpressionPayload.scala

```diff
@@ -126,48 +140,62 @@ class ExpressionPayload(record: GenericRecord,
+    val insertConditionAndAssignments =
+      ExpressionPayload.getEvaluator(insertConditionAndAssignmentsText.toString, writeSchema)
+    var resultRecordOpt: HOption[IndexedRecord] = null
+    for ((conditionEvaluator, assignmentEvaluator) <- insertConditionAndAssignments
+         if resultRecordOpt == null) {
+      val conditionVal = evaluate(conditionEvaluator, inputRecord).head.asInstanceOf[Boolean]
+      // If matched the insert condition then execute the assignment expressions to compute the
+      // result record. We will return the first matched record.
+      if (conditionVal) {
+        val results = evaluate(assignmentEvaluator, inputRecord)
+        resultRecordOpt = HOption.of(convertToRecord(results, writeSchema))
+      }
+    }
+    if (resultRecordOpt != null) {
+      resultRecordOpt
+    } else {
+      // If there is no condition matched, just filter this record.
+      // Here we return a IGNORE_RECORD, HoodieCreateHandle will not handle it.
+      HOption.of(HoodieWriteHandle.IGNORE_RECORD)
+    }
+  }
+
   override def getInsertValue(schema: Schema, properties: Properties): HOption[IndexedRecord] = {
     val incomingRecord = bytesToAvro(recordBytes, schema)
     if (isDeleteRecord(incomingRecord)) {
       HOption.empty[IndexedRecord]()
     } else {
-      val insertConditionAndAssignmentsText =
-        properties.get(ExpressionPayload.PAYLOAD_INSERT_CONDITION_AND_ASSIGNMENTS)
-      // Process insert
       val sqlTypedRecord = new SqlTypedRecord(incomingRecord)
-      // Get the evaluator for each condition and insert assignment.
-      initWriteSchemaIfNeed(properties)
-      val insertConditionAndAssignments =
-        ExpressionPayload.getEvaluator(insertConditionAndAssignmentsText.toString, writeSchema)
-      var resultRecordOpt: HOption[IndexedRecord] = null
-      for ((conditionEvaluator, assignmentEvaluator) <- insertConditionAndAssignments
-           if resultRecordOpt == null) {
-        val conditionVal = evaluate(conditionEvaluator, sqlTypedRecord).head.asInstanceOf[Boolean]
-        // If matched the insert condition then execute the assignment expressions to compute the
-        // result record. We will return the first matched record.
-        if (conditionVal) {
-          val results = evaluate(assignmentEvaluator, sqlTypedRecord)
-          resultRecordOpt = HOption.of(convertToRecord(results, writeSchema))
-        }
-      }
-
-      // Process delete for MOR
-      if (resultRecordOpt == null && isMORTable(properties)) {
-        val deleteConditionText = properties.get(ExpressionPayload.PAYLOAD_DELETE_CONDITION)
-        if (deleteConditionText != null) {
-          val deleteCondition = getEvaluator(deleteConditionText.toString, writeSchema).head._1
-          val deleteConditionVal = evaluate(deleteCondition, sqlTypedRecord).head.asInstanceOf[Boolean]
-          if (deleteConditionVal) {
-            resultRecordOpt = HOption.empty()
-          }
+      if (isMORTable(properties)) {
```

Comment:
Has moved the logical to the `processMatchedRecord` and `processNotMatchedRecord` now. The `processMatchedRecord` process the matched record, while the `processNotMatchedRecord` process the not matched record. This will be more clear.